<?xml version="1.0"?>
<launch>

  <!-- 下面的很多内容都是从 turtlebot3_slam.launch 文件抄过来的 -->
  <!-- Arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="slam_methods" default="gmapping" doc="slam type [gmapping, cartographer, hector, karto, frontier_exploration]"/>
  <arg name="configuration_basename" default="turtlebot3_lds_2d.lua"/>
  <arg name="open_rviz" default="true"/>
<include file="$(find turtlebot3_gazebo)/launch/turtlebot3_house.launch"/>
  <!-- TurtleBot3 -->
  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)" />
  </include>

  <!-- SLAM: Gmapping, Cartographer, Hector, Karto, Frontier_exploration, RTAB-Map -->
  <include file="$(find turtlebot3_slam)/launch/turtlebot3_karto.launch">
    <arg name="model" value="$(arg model)"/>
    <arg name="configuration_basename" value="$(arg configuration_basename)"/>
  </include>

  <!-- TODO 3.3.5 补充启动 rviz 的标签, 注意使用我们自己保存的 rviz 配置文件 -->
  <group if="$(arg open_rviz)"> 
    <node pkg="rviz" type="rviz" name="rviz" required="true"
          args="-d $(find octo_mapping)/rviz/point_cloud_mapping.rviz"/>
  </group>

  <!-- 启动 point_cloud_map 节点的标签, 同时完成参数服务器的设置 -->
  <node pkg="octo_mapping" type="point_cloud_map" name="point_cloud_map"  output="screen">
    <!-- 注意有无“/”的区别 -->
    <param name="/point_cloud_map/scale_factor_inv" value="0.5" />
    <param name="update_frequency" value="1" />
    <param name="distance_threshold" value="0.2" />
    <param name="angle_deg_threshold" value="10.0" />
    <param name="point_cloud_leaf_size" value="0.05" />
    <param name="point_cloud_mean_K" value="10" />
    <param name="point_cloud_filter_thres" value="0.2" />
  </node>
  <node pkg="octo_mapping" type="octo_map" name="octo_map"  output="screen"/>
<node pkg="turtlebot3_teleop" type="turtlebot3_teleop_key" name="turtlebot3_teleop_key" />
</launch>
